plugins/modem-manager/fu-qmi-pdc-updater.c

```diff
@@ -0,0 +1,433 @@
+				15, NULL, fu_qmi_pdc_updater_qmi_device_close_ready, ctx);
+}
+
+gboolean
+fu_qmi_pdc_updater_close (FuQmiPdcUpdater *self, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	CloseContext ctx = {
+		.mainloop = mainloop,
+		.qmi_device = g_steal_pointer (&self->qmi_device),
+		.qmi_client = g_steal_pointer (&self->qmi_client),
+	};
+
+	qmi_device_release_client (ctx.qmi_device, QMI_CLIENT (ctx.qmi_client),
+				   QMI_DEVICE_RELEASE_CLIENT_FLAGS_RELEASE_CID,
+				   5, NULL, fu_qmi_pdc_updater_qmi_device_release_client_ready, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* we should always have both device and client cleared, and optionally error set */
+
+	g_assert (ctx.qmi_device == NULL);
+	g_assert (ctx.qmi_client == NULL);
+
+	if (ctx.error != NULL) {
+		g_propagate_error (error, ctx.error);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+#define QMI_LOAD_CHUNK_SIZE 0x400
+
+typedef struct {
+	GMainLoop	*mainloop;
+	QmiClientPdc	*qmi_client;
+	GError		*error;
+	gulong		 indication_id;
+	guint		 timeout_id;
+	GBytes		*blob;
+	GArray		*digest;
+	gsize		 offset;
+	guint		 token;
+} WriteContext;
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-function"
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(QmiMessagePdcLoadConfigInput, qmi_message_pdc_load_config_input_unref)
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(QmiMessagePdcLoadConfigOutput, qmi_message_pdc_load_config_output_unref)
+#pragma clang diagnostic pop
+
+static void fu_qmi_pdc_updater_load_config (WriteContext *ctx);
+
+static gboolean
+fu_qmi_pdc_updater_load_config_timeout (gpointer user_data)
+{
+	WriteContext *ctx = user_data;
+
+	ctx->timeout_id = 0;
+	g_signal_handler_disconnect (ctx->qmi_client, ctx->indication_id);
+	ctx->indication_id = 0;
+
+	ctx->error = g_error_new_literal (G_IO_ERROR, G_IO_ERROR_FAILED,
+					  "couldn't load mcfg: timed out");
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_qmi_pdc_updater_load_config_indication (QmiClientPdc *client,
+					   QmiIndicationPdcLoadConfigOutput *output,
+					   WriteContext *ctx)
+{
+	gboolean frame_reset;
+	guint32 remaining_size;
+	guint16 error_code = 0;
+
+	g_source_remove (ctx->timeout_id);
+	ctx->timeout_id = 0;
+	g_signal_handler_disconnect (ctx->qmi_client, ctx->indication_id);
+	ctx->indication_id = 0;
+
+	if (!qmi_indication_pdc_load_config_output_get_indication_result (output, &error_code, &ctx->error)) {
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (error_code != 0) {
+		/* when a given mcfg file already exists in the device, an "invalid id" error is returned;
+		 * the error naming here is a bit off, as the same protocol error number is used both for
+		 * 'invalid id' and 'invalid qos id'
+		 */
+		if (error_code == QMI_PROTOCOL_ERROR_INVALID_QOS_ID) {
+			g_debug ("file already available in device");
+			g_main_loop_quit (ctx->mainloop);
+			return;
+		}
+
+		ctx->error = g_error_new (G_IO_ERROR, G_IO_ERROR_FAILED,
```

Comment:
`g_set_error?`

Reply:
just pushed a new fixup commit to handle this